### PR TITLE
CLI,  VSCode and LSP errors hyperlink to online docs

### PR DESCRIPTION
**Pull request by ahmafi**
Fix for #506 

- [x]  CLI
- [x]  VS Code extension
- [x]  LSP server

### CODE REVIEW

File: test/test-cli.py [hunk 1]
```diff
@@ -242,7 +242,7 @@ def test_config_error_for_multiple_js_files_is_printed_only_once(self) -> None:
                 encoding="utf-8",
             )
             self.assertEqual(result.returncode, 1)
-            self.assertEqual(result.stderr.count("E0164"), 1)
+            self.assertEqual(result.stderr.count("E0164"), 2)
```
Comment (strager) [inline test/test-cli.py:245]: This is confusing. We should instead do something like strip escape sequences in the test.

Must fix: Actually, the CLI shouldn't use escape sequences if the output is a pipe. We should only print escape sequences if the output is a terminal. (You can check if the output is a terminal using `isatty` on POSIX platforms.)

File: test/test-lsp-error-reporter.cpp [hunk 2]
```diff
@@ -62,6 +62,8 @@ TEST_F(test_lsp_error_reporter, big_int_literal_contains_decimal_point) {
             "BigInt literal contains decimal point");
   EXPECT_EQ(look_up(diagnostics, 0, "code"), "E0005");
   EXPECT_EQ(look_up(diagnostics, 0, "source"), "quick-lint-js");
+  EXPECT_EQ(look_up(diagnostics, 0, "codeDescription", "href"),
+            "https://quick-lint-js.com/errors/E0005");
```
Comment (strager) [inline test/test-lsp-error-reporter.cpp:66]: This link doesn't work yet. We should update the website first.

File: src/text-error-reporter.cpp [hunk 3]
```diff
@@ -63,7 +63,9 @@ void text_error_formatter::write_message_part(
 void text_error_formatter::write_after_message(std::string_view code,
                                                diagnostic_severity,
                                                const source_code_span &) {
-  this->output_ << " [" << code << "]\n";
+  this->output_ << " ["
+                << "\x1B]8;;https://quick-lint-js.com/errors/" << code
+                << "\x1B\\" << code << "\x1B]8;;\x1B\\]\n";
```
Comment (strager) [inline src/text-error-reporter.cpp:68]: Must fix: We shouldn't do this when writing to a pipe or a file. Use `isatty` to check if we are writing to a terminal.

It might be impossible to call `isatty` here since `this->output_` is an `std::ostream`, which might not be backed by a file descriptor. So perhaps we should pass the result of `isatty` to `text_error_reporter`'s constructor.

We might want to provide the option in the constructor anyway if we provide the user the option of enabling or disabling these escape codes. For example, GCC (a C++ compiler) allows the user to choose whether URLs are included for diagnostics using the `-fdiagnostics-urls` option: https://gcc.gnu.org/onlinedocs/gcc-11.2.0/gcc/Diagnostic-Message-Formatting-Options.html

File: plugin/vscode/test/vscode-tests.js [hunk 4]
```diff
@@ -52,9 +52,19 @@ for (let extension of [".js", ".mjs", ".cjs"]) {
...
           {
-            code: "E0034",
+            code: {
+              target: {
+                scheme: "https",
+                authority: "quick-lint-js.com",
+                path: "/errors/E0034",
+                query: "",
+                fragment: "",
+              },
```
Comment (strager) [inline plugin/vscode/test/vscode-tests.js:65]: Let's make `normalizeDiagnostics` convert the URI object into a string. This will let us simplify the assertion:
```suggestion
              target: "https://quick-lint-js.com/errors/E0034",
```

File: plugin/vscode/quick-lint-js/vscode-error-reporter.h [hunk 5]
```diff
@@ -60,7 +60,22 @@ class vscode_error_formatter
...
+
+    char error_path[14];
+    sprintf(error_path, "/errors/%s", code.data());
+
+    ::Napi::Object uri = ::Napi::Object::New(this->env_);
+    uri.Set("scheme", ::Napi::String::New(this->env_, "https"));
+    uri.Set("authority", ::Napi::String::New(this->env_, "quick-lint-js.com"));
+    uri.Set("path", ::Napi::String::New(this->env_, error_path));
+    uri.Set("fragment", ::Napi::String::New(this->env_, ""));
+    uri.Set("query", ::Napi::String::New(this->env_, ""));
```
Comment (strager) [inline plugin/vscode/quick-lint-js/vscode-error-reporter.h:72]: Must fix: The TypeScript definitions say that the `target` property must be an instance of `Uri`. A new object with the default prototype is not an instance of `Uri`. We should adhere to the documented specification:
```typescript
    export class Diagnostic {
        code?: string | number | {
            value: string | number;
            target: Uri;
        };

        // ...
      }
```

File: plugin/vscode/quick-lint-js/vscode-error-reporter.h [hunk 6]
```diff
@@ -60,7 +60,22 @@ class vscode_error_formatter
             this->env_, reinterpret_cast<const char*>(this->message_.data())),
         /*severity=*/severity,
     });
-    diag.Set("code", ::Napi::String::New(this->env_, code.data(), code.size()));
+
+    char error_path[14];
+    sprintf(error_path, "/errors/%s", code.data());
```
Comment (strager) [inline plugin/vscode/quick-lint-js/vscode-error-reporter.h:65]: Must fix: Use `std::snprintf`, not `sprintf`. Also, `QLJS_ASSERT` that `sprintf`/`snprintf` didn't fail and that the result is not truncated.

Or, be lazy and use `std::string` concatenation.

**Review by strager — CHANGES_REQUESTED** [review]

File: test/test-lsp-error-reporter.cpp (hunk 2 above)
Comment (ahmafi) [reply test/test-lsp-error-reporter.cpp:66]: Should I change the links to the fragment version? Or we update the website later?

File: src/text-error-reporter.cpp (hunk 3 above)
Comment (ahmafi) [reply src/text-error-reporter.cpp:68]: Should we do both?
1. User option for enabling or disabling escape codes
2. Check `stdout` with `isatty`

Comment (strager) [reply src/text-error-reporter.cpp:68]: > Should I do both?
>
> 1. User option for enabling or disabling escape codes
> 2. Check `stdout` with `isatty`

Yes. The default will be #2 ("auto"), and we allow the user to overwrite the decision with #1.

> Also should we write different tests for escaped characters?

In general, we should test what we want to work.

A "proper" test involves writing a terminal emulator. Not worth it. A "hacky" test involves string comparisons, in addition to manual testing, which is fine IMO.

File: test/test-lsp-error-reporter.cpp (hunk 2 above)
Comment (strager) [reply test/test-lsp-error-reporter.cpp:66]: > Should I change the links to the fragment version?

Yeah.

File: plugin/vscode/test/vscode-tests.js [hunk 7]
```diff
@@ -1192,7 +1230,14 @@ function normalizeDiagnostics(vscodeDiagnosticsOrURI) {
     vscodeDiagnostics = vscodeDiagnosticsOrURI;
   }
   return vscodeDiagnostics.map((diag) => ({
-    code: diag.code,
+    code: {
+      target: `${diag.code.target.scheme}://${diag.code.target.authority}${
+        diag.code.target.path
+      }${diag.code.target.query ? "?" + diag.code.target.query : ""}${
+        diag.code.target.fragment ? "#" + diag.code.target.fragment : ""
+      }`,
```
Comment (strager) [inline plugin/vscode/test/vscode-tests.js:1238]: Does `diag.code.target.toString()` not work?

File: src/main.cpp [hunk 8]
```diff
@@ -612,6 +614,11 @@ void print_help_message() {
   print_option("", "gnu-like (default if omitted)");
   print_option("", "vim-qflist-json");
   print_option("", "emacs-lisp");
+  print_option("--escape-errors=[WHEN]",
+               "Use escape sequences to show errors as hyperlinks.");
```
Comment (strager) [inline src/main.cpp:618]: Must fix: `--escape-errors` is a bad name. Let's call it `--diagnostic-hyperlinks` instead.

The fact that we use escape sequences is an implementation detail.

File: src/main.cpp [hunk 9]
```diff
@@ -612,6 +614,11 @@ void print_help_message() {
...
+  print_option("", "WHEN can be 'auto' (default if omitted), 'always',");
+  print_option("", "'never'. auto uses escape sequences when the standard");
+  print_option("", "error is a terminal");
```
Comment (strager) [inline src/main.cpp:621]: Nit: Detailed usage documentation belongs in the man page (`docs/cli.adoc`).

File: src/text-error-reporter.cpp [hunk 10]
```diff
@@ -23,16 +35,42 @@ void text_error_reporter::set_source(padded_string_view input,
...
   QLJS_ASSERT(this->locator_.has_value());
+
+  bool format_escape_errors = false;
+
+  switch (this->escape_errors_) {
+  case escape_errors::auto_:
+#if defined(_WIN32)
+    if (this->output_.rdbuf() == std::cerr.rdbuf() && _isatty(_fileno(stderr)))
+#else
+    if (this->output_.rdbuf() == std::cerr.rdbuf() && isatty(fileno(stderr)))
```
Comment (strager) [inline src/text-error-reporter.cpp:46]: Nit: No need for `fileno`:
```suggestion
    if (this->output_.rdbuf() == std::cerr.rdbuf() && isatty(STDIN_FILENO))
```

File: src/text-error-reporter.cpp [hunk 11]
```diff
@@ -23,16 +35,42 @@ void text_error_reporter::set_source(padded_string_view input,
...
+    else
+      format_escape_errors = false;
+    break;
+  case escape_errors::always:
+    format_escape_errors = true;
+    break;
+  case escape_errors::never:
+    format_escape_errors = false;
+    break;
+  }
```
Comment (strager) [inline src/text-error-reporter.cpp:58]: Must fix: This is very expensive to do every time `report` is called. Do this work in the constructor.

File: test/test-lsp-server.cpp [hunk 12]
```diff
@@ -271,8 +271,12 @@ TEST_F(test_linting_lsp_server, opening_document_lints) {
                         "end": {"line": 0, "character": 9}
                       },
                       "severity": 1,
-                        "message": "variable used before declaration: x"
+                      "message": "variable used before declaration: x",
+                      "code": "E0058",
+                      "codeDescription": {
+                        "href": "https://quick-lint-js.com/errors/#E0058"
                       }
```
Comment (strager) [inline test/test-lsp-server.cpp:278]: Nit: This test doesn't really need these. I'd undo the changes to this test.

**Review by strager — CHANGES_REQUESTED** [review]
See 'must fix' comments. Other than that, this PR looks good to me.

**Comment by ahmafi** [comment]
I have also tested CLI manually in windows and it works as expected.

File: src/text-error-reporter.cpp [hunk 13]
```diff
@@ -10,9 +11,40 @@
...
+  case escape_errors::always:
+    this->format_escape_errors_ = true;
+    break;
+  case escape_errors::never:
+    this->format_escape_errors_ = false;
+    break;
+  }
+}
+
+bool text_error_reporter::use_escape_if_auto() {
```
Comment (strager) [inline src/text-error-reporter.cpp:41]: Nit: Maybe `output_supports_escapes` is a better name?

File: src/text-error-reporter.cpp [hunk 14]
```diff
@@ -23,16 +55,22 @@ void text_error_reporter::set_source(padded_string_view input,
...
   QLJS_ASSERT(this->file_path_);
   QLJS_ASSERT(this->locator_.has_value());
-  text_error_formatter formatter(/*output=*/this->output_,
-                                 /*file_path=*/this->file_path_,
-                                 /*locator=*/*this->locator_);
+  text_error_formatter formatter(
+      /*output=*/this->output_,
+      /*file_path=*/this->file_path_,
+      /*locator=*/*this->locator_,
+      /*format_escape_errors*/ this->format_escape_errors_);
```
Comment (strager) [inline src/text-error-reporter.cpp:62]: Nit:
```suggestion
      /*format_escape_errors=*/this->format_escape_errors_);
```

File: docs/cli.adoc [hunk 15]
```diff
@@ -48,6 +48,19 @@ This command has two modes:
 Incompatible with *--lsp-server*.
 --
 
+*--diagnostic-hyperlinks*=_when_::
+  Control whether the errors are using escape sequences to show errors as hyperlinks or not.
+  _when_ is one of the following:
++
+--
+- *auto* (default): shows errors as hyperlinks only if the error output is a terminal. On windows
+it behaves like 'never' because windows default terminals don't support escape sequences.
```
Comment (strager) [inline docs/cli.adoc:57]: Nit: One line per sentence:
```suggestion
- *auto* (default): shows errors as hyperlinks only if the error output is a terminal.
On windows it behaves like 'never' because windows default terminals don't support escape sequences.
```

Comment (strager) [inline docs/cli.adoc:57]: Minor: I wouldn't mention why it behaves line `never` on Windows.
```suggestion
- *auto* (default): shows errors as hyperlinks only if the error output is a terminal. On windows
it behaves like 'never'.
```

File: docs/cli.adoc [hunk 16]
```diff
@@ -48,6 +48,19 @@ This command has two modes:
 Incompatible with *--lsp-server*.
 --
 
+*--diagnostic-hyperlinks*=_when_::
+  Control whether the errors are using escape sequences to show errors as hyperlinks or not.
+  _when_ is one of the following:
++
+--
+- *auto* (default): shows errors as hyperlinks only if the error output is a terminal. On windows
```
Comment (strager) [inline docs/cli.adoc:56]: Nit: Windows is capitalized:
```suggestion
- *auto* (default): shows errors as hyperlinks only if the error output is a terminal. On Windows
```

File: docs/cli.adoc [hunk 17]
```diff
@@ -48,6 +48,19 @@ This command has two modes:
 Incompatible with *--lsp-server*.
 --
 
+*--diagnostic-hyperlinks*=_when_::
+  Control whether the errors are using escape sequences to show errors as hyperlinks or not.
```
Comment (strager) [inline docs/cli.adoc:52]: "show errors as hyperlinks" is weird phrasing IMO. Hyperlink is a verb (as well as a noun), so we can say "hyperlink errors".

But we're not really hyperlinking the errors; we're hyperlinking the error codes. So maybe write this:
```suggestion
  Control whether to hyperlink error codes or not.
```

File: plugin/vscode/quick-lint-js/vscode-error-reporter.h [hunk 18]
```diff
@@ -60,7 +60,20 @@ class vscode_error_formatter
...
     });
-    diag.Set("code", ::Napi::String::New(this->env_, code.data(), code.size()));
+
+    ::Napi::Value uri = this->vscode_->uri_class.New(
+        {/*scheme*/ ::Napi::String::New(this->env_, "https"),
+         /*authority*/ ::Napi::String::New(this->env_, "quick-lint-js.com"),
+         /*path*/ ::Napi::String::New(this->env_, "/errors/"),
+         /*query*/ ::Napi::String::New(this->env_, ""),
+         /*fragment*/
+         ::Napi::String::New(this->env_, code.data(), code.size())});
```
Comment (strager) [inline plugin/vscode/quick-lint-js/vscode-error-reporter.h:70]: Nit:
```suggestion
        {/*scheme=*/::Napi::String::New(this->env_, "https"),
         /*authority=*/::Napi::String::New(this->env_, "quick-lint-js.com"),
         /*path=*/::Napi::String::New(this->env_, "/errors/"),
         /*query=*/::Napi::String::New(this->env_, ""),
         /*fragment=*/
         ::Napi::String::New(this->env_, code.data(), code.size())});
```

File: docs/cli.adoc (hunk 17 above)
Comment (strager) [inline docs/cli.adoc:52]: I'd also mention that this option is only used if `--output-format=gnu-like`.

**Review by strager — APPROVED** [review]

File: src/main.cpp [hunk 19]
```diff
@@ -612,6 +614,11 @@ void print_help_message() {
   print_option("", "gnu-like (default if omitted)");
   print_option("", "vim-qflist-json");
   print_option("", "emacs-lisp");
+  print_option("--diagnostic-hyperlinks=[WHEN]",
+               "Control if errors are shown as hyperlink. WHEN is one of:");
+  print_option("", "auto (default if omitted)");
```
[the remark this answers is not in the document]
Comment (ahmafi) [reply src/main.cpp:619]: Should I remove "if omitted" from "--output-format" too? "Format" can't be omitted there too.

[the remark this answers is not in the document]
Comment (strager) [reply src/main.cpp:619]: > Should I remove "if omitted" from "--output-format" too? "Format" can't be omitted there too.

Yeah. It's there in `--output-format` by mistake.

**Comment by strager** [comment]
Looks awesome. Thanks for improving the completions!

Can you write a commit message for this PR? I plan to squash your commits into one. (We can land separate commits if you really want.)
